WordPress/WordPressApi/WordPressComRestApi.swift

```diff
@@ -272,6 +247,163 @@ public class WordPressComRestApi: NSObject
+            progress.totalUnitCount = contentLength + 1
+        }
+        progress.cancellationHandler = {
+            task.cancel()
+        }
+        ongoingProgress[task] = progress
+    }
+
+    //MARK: - Response handling
+
+    private func handleResponseWithData(originalData: NSData?, urlResponse:NSURLResponse?, error: NSError?) throws -> AnyObject {
+        guard let data = originalData,
+            let httpResponse = urlResponse as? NSHTTPURLResponse,
+            let contentType = httpResponse.allHeaderFields["Content-Type"] as? String
+            where error == nil &&  contentType.hasPrefix("application/json") else {
+                if let unwrappedError = error {
+                    throw convertError(unwrappedError, data: originalData)
+                } else {
+                    throw convertError(WordPressComRestApiError.ResponseSerializationFailed as NSError, data: originalData)
+                }
+        }
+
+        let jsonObject: AnyObject
+
+        do {
+            jsonObject = try NSJSONSerialization.JSONObjectWithData(data, options: NSJSONReadingOptions())
+        } catch {
+            throw convertError(WordPressComRestApiError.ResponseSerializationFailed as NSError, data: originalData)
+        }
+        if (200..<300).contains(httpResponse.statusCode) {
+            return jsonObject
+        }
+        if (400...500).contains(httpResponse.statusCode) {
+            throw handleErrorResponseObject(jsonObject, originalData: data)
+        }
+        throw convertError(WordPressComRestApiError.ResponseSerializationFailed as NSError, data: originalData)
+    }
+
+    //MARK: - Error Handling
+
+    private func handleErrorResponseObject(responseObject: AnyObject, originalData:NSData) -> NSError {
+        guard let responseDictionary = responseObject as? [String:AnyObject] else {
+            return convertError(WordPressComRestApiError.ResponseSerializationFailed as NSError, data: originalData)
+        }
+        var errorDictionary:AnyObject? = responseDictionary
+        if let errorArray = responseDictionary["errors"] as? [AnyObject] where errorArray.count > 0 {
+            errorDictionary = errorArray.first
+        }
+        guard let errorEntry = errorDictionary as? [String:AnyObject],
+            let errorCode = errorEntry["error"] as? String,
+            let errorDescription = errorEntry["message"] as? String
+            else {
+                return convertError(WordPressComRestApiError.ResponseSerializationFailed as NSError, data: originalData)
+        }
+
+        let errorsMap = [
+            "invalid_input" : WordPressComRestApiError.InvalidInput,
+            "invalid_token" : WordPressComRestApiError.InvalidToken,
+            "authorization_required" : WordPressComRestApiError.AuthorizationRequired,
+            "upload_error" : WordPressComRestApiError.UploadFailed,
+            "unauthorized" : WordPressComRestApiError.AuthorizationRequired
+        ]
```

Comment:
The issue is that if I change the enum to string it will mess up the auto translation of the error code to Objective-C.
And because we still have some objective-c code checking for this errors we don't want to do that.